### PR TITLE
Notificationlogic

### PR DESCRIPTION
**Pull request by Steindt**
Backend and updated frontend for web notifications.

Currently only linked to creating news, updating news and pinging as these are the ones implemented so far.

To test for notifications against yourself, uncomment line 95 in _src/lib/utils/notifications/index.ts_.

Part of #30 and completes #168.

### CODE REVIEW

File: src/database/prisma/schema.prisma
Comment (Macludde) [inline src/database/prisma/schema.prisma:510]: Does this mean these rows have to be updated whenever a new way to send notficiations is added? Perhaps we should just set "create" to "_" or whatever "logged in" is?

File: src/database/prisma/schema.prisma
Comment (Macludde) [inline src/database/prisma/schema.prisma:578]: Should this include the above check as well, ie that the user is one of the included members in the ping?

File: src/lib/utils/notifications/index.ts
Comment (Macludde) [inline src/lib/utils/notifications/index.ts:158]: remove?

**Review by Macludde — COMMENTED** [review]
I'm still reviewing this, here are some things thus far. When I try to start it locally, I get this error:

```
{
  name: 'PrismaClientKnownRequestError',
  code: 'P2023',
  clientVersion: '5.7.1',
  meta: {
    modelName: 'Notification',
    message: 'Error creating UUID, invalid character: expected an optional prefix of `urn:uuid:` followed by [0-9a-fA-F-], found `n` at 2'
  },
  internalStack: 'PrismaClientKnownRequestError: \n' +
    'Invalid `prisma.notification.findMany()` invocation:\n' +
    '\n' +
    '\n' +
    'Inconsistent column data: Error creating UUID, invalid character: expected an optional prefix of `urn:uuid:` followed by [0-9a-fA-F-], found `n` at 2\n' +
    '    at si.handleRequestError (/Users/ludvigsvedberg/Programming/dsek/web/node_modules/.pnpm/@prisma+client@5.7.1_prisma@5.7.1/node_modules/@prisma/client/runtime/library.js:125:6817)\n' +
    '    at si.handleAndLogRequestError (/Users/ludvigsvedberg/Programming/dsek/web/node_modules/.pnpm/@prisma+client@5.7.1_prisma@5.7.1/node_modules/@prisma/client/runtime/library.js:125:6151)\n' +
    '    at si.request (/Users/ludvigsvedberg/Programming/dsek/web/node_modules/.pnpm/@prisma+client@5.7.1_prisma@5.7.1/node_modules/@prisma/client/runtime/library.js:125:5859)\n' +
    '    at async l (/Users/ludvigsvedberg/Programming/dsek/web/node_modules/.pnpm/@prisma+client@5.7.1_prisma@5.7.1/node_modules/@prisma/client/runtime/library.js:130:10025)'
}
```

**Comment by Macludde** [comment]
I've found some other smaller issues, but I think it would be easier for me to change them locally, push my commits and then you can have a look at them @Steindt

File: src/lib/utils/notifications/index.ts
Comment (Macludde) [inline src/lib/utils/notifications/index.ts:12]: I renamed this to be more clear. `ARTICLE_REQUEST_UPDATE` (as it's now called) is meant to be a notification you get when an update has been made to your active article request (not yet built in this version). Such as if it has been approved or rejected.

File: src/lib/components/Notification.svelte
Comment (Macludde) [inline src/lib/components/Notification.svelte:17]: We use superForm in this project, and thus we don't use the bultin `enhance` method, feel free to check out how I solved it with this setup for future PRs

File: src/lib/components/Notification.svelte
Comment (Macludde) [inline src/lib/components/Notification.svelte:10]: Instead of manually declaring the `Notification` object, use the type export from `@prisma/client`. And even better, if you don't use all properties, use `Pick<...>` to select only the properties you actually use. This is nice in the future if we want to do optimization and at a glance see which properties we need to extract from the DB for each page.

File: src/routes/(app)/read-notifications/+server.ts
Comment (Macludde) [inline src/routes/(app)/read-notifications/+server.ts:6]: So I re-built how "read notifications" is done. From an abstract perspective, marking notifications as read is not an explicit user actions, it is done when they click on the bell icon. Previously, this bell icon was within a `<form>` tag, and when clicked, a form would be submitted.

This worked because your browser has javascript enabled, and what superforms (or SvelteKit for that matter) does when javascript is enabled, is that they stop the form submission and instead just sends a fetch request to the server. Regular form requests always lead to a page load. This means notifications would not work for people without JS enabled. Even if those users are few, just semantically this did not make to be a form.

Instead I built a `PUT` endpoint which is called when the bell icon is clicked, marking all notifications as read.

Comment (Macludde) [reply src/routes/(app)/read-notifications/+server.ts:6]: Of course, this means non-JS users can not mark their notifications as read. That's a bummer and if necessary, we might want to add some support for that in the future (like a "mark as read" button which is actually a form)

File: src/routes/+layout.server.ts
Comment (Macludde) [inline src/routes/+layout.server.ts:26]: Superform stuff, if you want you can read more on superforms.rocks

File: src/routes/(app)/NotificationBell.svelte
Comment (Macludde) [inline src/routes/(app)/NotificationBell.svelte:16]: This was previously a for loop which mutated an array. This meant reactivity didn't work. In svelte, as opposed to say react, a components script code is only run once, not on every "re-render". Re-renders in svelte work by finding dependencies, for example:
- If you declare a variable with let, it will keep track everywhere that is user
- Variables declared as `$: varName = ...` are called derived and svelte will keep track of which variables they use, the derived property will be updated whenever one of its dependencies changes.

**Comment by Macludde** [comment]
@Steindt I pushed some changes, I also added comments above with the biggest changes.

In summary, the biggest things were:
- We use superForms for form validation, and thus "regular" svelte-kit forms (with the `enhance` imported from svelte) should not be used, instead the `+page.server.ts` should return form data which is used on the client
- `readNotifications` should not be a form action, but instead just an api endpoint as its not an explicit action a user takes. This works in general but breaks for users who do not have JS enabled. A big thing with svelte is the ability for it to work for users who turn off JS (most if not almost all logic is server side). If a user has JS enabled (which are most users) it adds some flair to make the page act more like a SPA (use:enhance for example)
- Import types, use the types from `@prisma/client` instead of declaring the object.

Also, please use camelCase for methods or form actions (the form actions were called `deletenotification` and `readnotification`)

But otherwise, I like what you have done and it worked very well. I'll take a look at pings as well before approving

**Comment by Macludde** [comment]
I would say this is finished now and ready to be merged

**Review by Macludde — APPROVED** [review]

File: src/routes/(app)/read-notifications/+server.ts
Comment (Steindt) [reply src/routes/(app)/read-notifications/+server.ts:6]: Good point!

File: src/lib/utils/notifications/index.ts
Comment (danieladugyan) [inline src/lib/utils/notifications/index.ts:134]:
```suggestion
  const databaseResult = await prisma.notification.createMany({
    data: receivingMembers.map(({ id: memberId }) => ({
      title,
      message,
      type,
      link,
      memberId,
      fromAuthorId: notificationAuthor?.id,
    })),
  });
```

File: src/lib/zod/schemas.ts
Comment (danieladugyan) [inline src/lib/zod/schemas.ts:5]: This is very odd to me. Zod is schema validation library. What's the point of validating an empty object?

File: src/routes/(app)/members/[studentId]/+page.server.ts
Comment (danieladugyan) [inline src/routes/(app)/members/[studentId]/+page.server.ts:78]:
```suggestion
                  toMemberId: user.memberId,
```

File: src/routes/(app)/members/[studentId]/+page.server.ts
Comment (danieladugyan) [inline src/routes/(app)/members/[studentId]/+page.server.ts:81]:
```suggestion
                  fromMemberId: user.memberId,
```

File: src/routes/(app)/members/[studentId]/+page.server.ts
Comment (danieladugyan) [inline src/routes/(app)/members/[studentId]/+page.server.ts:94]: Delete unreachable code.

File: src/routes/(app)/members/[studentId]/pings.ts
Comment (danieladugyan) [inline src/routes/(app)/members/[studentId]/pings.ts:75]:
```suggestion
        ? { id: { equals: member.memberId } }
```

File: src/routes/(app)/members/[studentId]/pings.ts
Comment (danieladugyan) [inline src/routes/(app)/members/[studentId]/pings.ts:77]:
```suggestion
            studentId: member.studentId,
```

File: src/routes/+layout.server.ts
Comment (danieladugyan) [inline src/routes/+layout.server.ts:1]: Notification code shouldn't be in this top-level layout file.

File: src/routes/(app)/Navbar.svelte
Comment (danieladugyan) [inline src/routes/(app)/Navbar.svelte:15]: Maybe we should get notifications from $page.data.notifications instead? I'm not sure what's best.

File: src/routes/(app)/+page.server.ts
Comment (danieladugyan) [inline src/routes/(app)/+page.server.ts:93]: I can't help but feel that noitifications don't really belong here. They don't really have anything to do with this page.

**Review by danieladugyan — CHANGES_REQUESTED** [review]
Looks awesome overall! Sorry for the slow review.

Parts of the code are sometimes a bit hard to understand.  I'm especially concerned with how we want to handle reading and dispatching notifications across the app. Should we use an API-route for reading notifications? Should we use a form action for deleting notifications? It's not obvious which pattern is easier to understand and maintain, but maybe we should decide on one of them.

and I'm not good enough at SvelteKit to know how to solve some of the issues I raised. Page-wide data/actions tend to creep

**Comment by danieladugyan** [comment]
Tagging you @Macludde since I think some of my comments pertain to your changes.

File: src/routes/(app)/members/[studentId]/+page.server.ts
Comment (Steindt) [reply src/routes/(app)/members/[studentId]/+page.server.ts:94]: Out of place after merge? Moving it to to be included in return

**Comment by Steindt** [comment]
@danieladugyan and I decided to use forms, instead of API calls, when reading notifications to remain consistent in the code.

File: src/routes/(app)/+page.server.ts
Comment (Steindt) [reply src/routes/(app)/+page.server.ts:93]: Has been moved to separate page

File: src/routes/(app)/Navbar.svelte
Comment (Macludde) [reply src/routes/(app)/Navbar.svelte]: I agree

**Comment by Macludde** [comment]
@Steindt we cannot have the bell as a form action. It does not work for users without js enabled at all, because the form action will result in a page reload, which then would close the dropdown.

**Comment by Macludde** [comment]
I changed it so it now uses forms, but also works for non-JS users.

Basically, it no longer automatically reads all notifications when you click the bell icon. At the bottom, there is a "mark all as read button", and if you visit a notification, that notification is marked unread.

File: src/lib/zod/schemas.ts
Comment (Macludde) [reply src/lib/zod/schemas.ts:5]: Superforms require a zod schema, so for a form where no form data is expected, I think using an empty schema makes sense

**Comment by Macludde** [comment]
@danieladugyan would you like to take a quick look and approve if it looks good? I'm somewhat happy with the solution now. It completely uses forms, using a "hidden form" to automatically mark a notification as read. 

I changed the logic, so instead of marking all notifications as read when opening notifications, it now only marks them when read if the user visits the link they are pointing to (for example by clicking on them). I think this makes more sense, and also makes the app a lot easier, because it only has to open the relevant link and the notifications will sort themselves out.

The notification code has been moved to a seperate /notifications page by @Steindt, this moves away some code but does make `/notifications` a "valid" endpoint for the page. Currently, this page is never visited as a JS-enabled user, and for non-JS users it will redirect them back to where they came from.

**Pull request merged**
